File: android/src/main/kotlin/com/situm/flutter/ar/situm_ar/scene/ARQuality.kt
package com.situm.flutter.ar.situm_ar.scene

import dev.romainguy.kotlin.math.pow
import es.situm.sdk.model.location.Location
import io.github.sceneview.math.Position
import io.github.sceneview.math.Rotation
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt


const val BUFFER_SIZE = 15

const val DEFAULT_REFRESH_THRESHOLD = 0.2
const val CONSTANT_QUALITY_DECREASE_RATE = 0.005        // Every iteration always decreases this
const val QUALITY_THRESHOLD_DECREASE_RATE = 0.03        // when low quality, threshold decreases faster

data class RefreshThreshold(var value: Double, var timestamp: Long) {
    override fun toString(): String {
        return "RefreshThreshold(value: $value, timestamp: $timestamp)"
    }
}


class ARQuality {

    private var currentRefreshThreshold: RefreshThreshold = RefreshThreshold(0.2,0)
    private var dynamicRefreshThreshold: RefreshThreshold = RefreshThreshold(0.2,0)

    private var quality: Double = 0.0

    private var odometriesDistanceConf: Double = 0.0
    private var situmConf: Double = 0.0
    private var arConf: Double= 0.0
    private var arDisplacementConf: Double= 0.0
    private var situmDisplacementConf: Double = 0.0

    private var arLocationBuffer: MutableList<LocationCoordinates> = mutableListOf()
    private var situmLocationBuffer: MutableList<LocationCoordinates> = mutableListOf()


    fun updateARLocation(worldPosition: Position, worldRotation: Rotation) {
        arLocationBuffer.add(LocationCoordinates(worldPosition.x.toDouble(),
            worldPosition.z.toDouble(), worldRotation.y.toDouble(), System.currentTimeMillis()))  //TODO: Check rotation
        if (arLocationBuffer.size> BUFFER_SIZE){
            arLocationBuffer.removeAt(0)
        }
    }

    fun updateSitumLocation(location: Location){
        if (situmLocationBuffer.isNotEmpty() && situmLocationBuffer.last().floorIdentifier != location.floorIdentifier){
            situmLocationBuffer.clear()
            arLocationBuffer.clear()
            resetThreshold()
        }
        situmLocationBuffer.add(LocationCoordinates(location.cartesianCoordinate.x, location.cartesianCoordinate.y,location.cartesianBearing.degrees(),System.currentTimeMillis(),location.floorIdentifier,
            location.accuracy.toLong(), location.hasBearing() ))
        if (situmLocationBuffer.size> BUFFER_SIZE){
            situmLocationBuffer.removeAt(0)
        }
    }


    fun hasToResetWorld():Boolean{
        updateConfidence()
        return checkIfHasToRefreshAndUpdateThreshold(quality,arConf, situmConf)
    }
    fun updateConfidence() {
        if (situmLocationBuffer.isEmpty() || arLocationBuffer.isEmpty()){
            arConf = 0.0
            situmConf = 0.0
            quality = 0.0
            return
        }
        val  totalDisplacementSitum = computeTotalDisplacement(situmLocationBuffer)
        val  totalDisplacementAR = computeTotalDisplacement(arLocationBuffer)

        val odometriesDistance = estimateOdometriesMatch(arLocationBuffer, situmLocationBuffer)

        situmDisplacementConf = totalDisplacementConf(totalDisplacementSitum)
        arDisplacementConf = totalDisplacementConf(totalDisplacementAR)
        arConf = estimateArConf()
        situmConf = estimateSitumConf()
        odometriesDistanceConf = odometriesDifferenceConf(odometriesDistance)

        quality = situmDisplacementConf * arDisplacementConf * arConf * situmConf * odometriesDistanceConf
    }


    private fun estimateOdometriesMatch(arLocationBuffer: MutableList<LocationCoordinates>, situmLocationBuffer: MutableList<LocationCoordinates>): Double {
        var transformedARTrajectory = transformTrajectory(arLocationBuffer)
        var transformedSitumTrajectory = transformTrajectory(situmLocationBuffer)
        val distance = transformedARTrajectory.last().distanceTo(transformedSitumTrajectory.last())

        return distance;

    }

    fun transformTrajectory(arLocationBuffer: MutableList<LocationCoordinates>): List<LocationCoordinates> {
        if (arLocationBuffer.isEmpty()) return emptyList()

        // Translate to origin
        val origin = arLocationBuffer[0]
        val translatedTrajectory = arLocationBuffer.map { loc -> loc - origin }
        if (translatedTrajectory.size == 1) return translatedTrajectory

        // Find a miminum displacement
        var distance = 0.0
        var index = 1
        while (index < translatedTrajectory.size) {
            distance = translatedTrajectory[0].distanceTo(translatedTrajectory[index])
            if (distance > 2) break
            index++
        }
        if (index >= translatedTrajectory.size) {
            return translatedTrajectory
        }

        // Calculate rotation
        val firstVector = translatedTrajectory[index]
        val angle = atan2(firstVector.y, firstVector.x)

        // Apply rotation
        val alignedTrajectory = translatedTrajectory.map { loc -> loc.rotate(-angle) }

        return alignedTrajectory
    }


    fun computeTotalDisplacement(coordinates: List<LocationCoordinates>): Double {
        if (coordinates.size < 2) return 0.0
        return coordinates.first().distanceTo(coordinates.last())
    }


    fun estimateArConf(): Double {
        val requiredPositions = 10
        val maxConfidence = 1.0
        var numOkPositions = 0

        // Ckeck last 10 positions
        var confidence = maxConfidence
        for (i in arLocationBuffer.size - 1 downTo maxOf(arLocationBuffer.size - requiredPositions, 0)) {
            // Si no hay AR, se congela, recibimos el último valor nuevamente.      // TODO: Chek if this continues to be true with the new library
            if ((arLocationBuffer[i].x == 0.0 && arLocationBuffer[i].y == 0.0) ||
                i < 1 ||
                (arLocationBuffer[i].y == arLocationBuffer[i - 1].y && arLocationBuffer[i].x == arLocationBuffer[i - 1].x)
            ) {
                break
            } else {
                numOkPositions++
            }
        }
        confidence = (numOkPositions.toDouble() / requiredPositions) * maxConfidence
        return confidence
    }


    fun estimateSitumConf(): Double {
        val requiredPositions = 10
        val maxConfidence = 1.0
        var numOkPositions = 0
        var confidence = maxConfidence

        for (i in situmLocationBuffer.size - 1 downTo maxOf(situmLocationBuffer.size - requiredPositions, 0)) {
            if ((situmLocationBuffer[i].accuracy > 5 && !situmLocationBuffer[i].hasBearing) || i < 0) {
                break
            } else {
                numOkPositions++
            }
        }

        confidence = (numOkPositions.toDouble() / requiredPositions) * maxConfidence
        return confidence
    }

    fun totalDisplacementConf(distance: Double): Double {
        val minDistanceThreshold = 10.0
        return if (distance > minDistanceThreshold) {
            1.0
        } else {
            distance / minDistanceThreshold
        }
    }
    fun odometriesDifferenceConf(difference: Double): Double {
        val diffThreshold = 10.0
        return if (difference > diffThreshold) {
            0.0
        } else {
            1.0 - difference / diffThreshold
        }
    }

    fun checkIfHasToRefreshAndUpdateThreshold(conf: Double, arConf: Double, situmConf: Double): Boolean {
        val currentTimestamp = System.currentTimeMillis()

        // Si la confianza de AR o Situm es menor que 0.8, reiniciar el umbral y devolver true
        if (arConf < 0.8 || situmConf < 0.8) {
            resetThreshold()
            return true
        }

        // Reducir siempre el umbral si es necesario
        if (currentRefreshThreshold.value > 0.20 &&
            currentTimestamp - currentRefreshThreshold.timestamp > 1000) {
            currentRefreshThreshold.value = currentRefreshThreshold.value - CONSTANT_QUALITY_DECREASE_RATE
        }

        // Si la confianza es mayor que el umbral actual + 0.2, actualizar el umbral y devolver true
        if (conf > currentRefreshThreshold.value + 0.2) {
            currentRefreshThreshold.value = conf
            currentRefreshThreshold.timestamp = currentTimestamp
            dynamicRefreshThreshold = currentRefreshThreshold
            return true
        }
        // Si la confianza es menor que el umbral actual y ha pasado más de 1 segundo, disminuir el umbral
        else if (conf < currentRefreshThreshold.value &&
            currentTimestamp - currentRefreshThreshold.timestamp > 1000 &&
            currentRefreshThreshold.value > 0.20) {
            currentRefreshThreshold.value = currentRefreshThreshold.value - QUALITY_THRESHOLD_DECREASE_RATE
            currentRefreshThreshold.timestamp = currentTimestamp
            dynamicRefreshThreshold = currentRefreshThreshold
            return false
        }
        // Si ha pasado más de 30 segundos, actualizar el umbral con el valor de confianza o 0.2
        else if (currentTimestamp - currentRefreshThreshold.timestamp > 30000) {
            currentRefreshThreshold.value = if (conf > 0.2) conf else 0.2
            currentRefreshThreshold.timestamp = currentTimestamp
            dynamicRefreshThreshold = currentRefreshThreshold
            return true
        }

        return false
    }

    private fun resetThreshold() {
        currentRefreshThreshold.timestamp = System.currentTimeMillis()
        currentRefreshThreshold.value = DEFAULT_REFRESH_THRESHOLD
        dynamicRefreshThreshold = currentRefreshThreshold

    }


}

File: ios/Classes/ARView/ARQuality.swift
import Foundation
import SceneKit
import CoreLocation
import SitumSDK

let BUFFER_SIZE = 15

struct RefreshThreshold {
    var value: Double
    var timestamp: TimeInterval

    func description() -> String {
        return "RefreshThreshold(value: \(value), timestamp: \(timestamp))"
    }
}

class ARQuality {

    private var currentRefreshThreshold = RefreshThreshold(value: 0.2, timestamp: 0)
    private var dynamicRefreshThreshold = RefreshThreshold(value: 0.2, timestamp: 0)

    private var quality: Double = 0.0

    private var odometriesDistanceConf: Double = 0.0
    private var situmConf: Double = 0.0
    private var arConf: Double = 0.0
    private var arDisplacementConf: Double = 0.0
    private var situmDisplacementConf: Double = 0.0

    private var arLocationBuffer: [LocationCoordinates] = []
    private var situmLocationBuffer: [LocationCoordinates] = []
    
    var CONSTANT_QUALITY_DECREASE_RATE = 0.005
    var QUALITY_THRESHOLD_DECREASE_RATE = 0.03

    func updateARLocation(worldPosition: SCNVector3, worldRotation: SCNQuaternion) {
        let currentTime = Date().timeIntervalSince1970 * 1000 // Esto ya es Double (TimeInterval)
        
        arLocationBuffer.append(LocationCoordinates(x: Double(worldPosition.x),
                                                    y: Double(worldPosition.z),
                                                    yaw: Double(worldRotation.y),
                                                    timestamp: currentTime)) // currentTime ya es TimeInterval

        if arLocationBuffer.count > BUFFER_SIZE {
            arLocationBuffer.removeFirst()
        }
    }

    func setQualityDecrease(qualityDecrease: Float){
        self.CONSTANT_QUALITY_DECREASE_RATE = Double(qualityDecrease)        
    }
    func setThresholdDecrease(thresholdDecrease: Float){
        self.QUALITY_THRESHOLD_DECREASE_RATE = Double(thresholdDecrease)
    }


    func updateSitumLocation(location: SITLocation) {
        if !situmLocationBuffer.isEmpty, situmLocationBuffer.last?.floorIdentifier != location.position.floorIdentifier {
            situmLocationBuffer.removeAll()
            arLocationBuffer.removeAll()
            // TODO: RESET Threshold
        }
        
        // Desempaquetar de forma segura los valores opcionales
        if let cartesianCoordinate = location.position.cartesianCoordinate {
            let currentTime = Date().timeIntervalSince1970 * 1000
            
            situmLocationBuffer.append(LocationCoordinates(
                x: Double(cartesianCoordinate.x),
                y: Double(cartesianCoordinate.y),
                yaw: Double(location.cartesianBearing.degrees()),
                timestamp: currentTime,
                floorIdentifier: location.position.floorIdentifier ?? "",
                accuracy: Int64(location.accuracy),
                hasBearing: location.hasBearing()
            ))
            
            if situmLocationBuffer.count > BUFFER_SIZE {
                situmLocationBuffer.removeFirst()
            }
        } else {
            print("Error: cartesianCoordinate es nil")
        }
    }


    func hasToResetWorld() -> Bool {
        updateConfidence()
        return checkIfHasToRefreshAndUpdateThreshold(conf: quality, arConf: arConf, situmConf: situmConf)
    }

    func updateConfidence() {
        guard !situmLocationBuffer.isEmpty, !arLocationBuffer.isEmpty else {
            arConf = 0.0
            situmConf = 0.0
            quality = 0.0
            return
        }

        let totalDisplacementSitum = computeTotalDisplacement(coordinates: situmLocationBuffer)
        let totalDisplacementAR = computeTotalDisplacement(coordinates: arLocationBuffer)
        let odometriesDistance = estimateOdometriesMatch(arLocationBuffer: arLocationBuffer, situmLocationBuffer: situmLocationBuffer)

        situmDisplacementConf = totalDisplacementConf(distance: totalDisplacementSitum)
        arDisplacementConf = totalDisplacementConf(distance: totalDisplacementAR)
        arConf = estimateArConf()
        situmConf = estimateSitumConf()
        odometriesDistanceConf = odometriesDifferenceConf(difference: odometriesDistance)

        quality = situmDisplacementConf * arDisplacementConf * arConf * situmConf * odometriesDistanceConf
    }

    private func estimateOdometriesMatch(arLocationBuffer: [LocationCoordinates], situmLocationBuffer: [LocationCoordinates]) -> Double {
        let transformedARTrajectory = transformTrajectory(trajectory: arLocationBuffer)
        let transformedSitumTrajectory = transformTrajectory(trajectory: situmLocationBuffer)
        
        // Llamada a distanceTo sin el label `other`
        return transformedARTrajectory.last!.distanceTo(transformedSitumTrajectory.last!)
    }


    func transformTrajectory(trajectory: [LocationCoordinates]) -> [LocationCoordinates] {
        guard !trajectory.isEmpty else { return [] }

        // Translate to origin
        let origin = trajectory[0]
        let translatedTrajectory = trajectory.map { $0 - origin }
        if translatedTrajectory.count == 1 { return translatedTrajectory }

        // Find minimum displacement
        var distance = 0.0
        var index = 1
        while index < translatedTrajectory.count {
            // Llamada sin 'other:'
            distance = translatedTrajectory[0].distanceTo(translatedTrajectory[index])
            if distance > 2 { break }
            index += 1
        }

        guard index < translatedTrajectory.count else { return translatedTrajectory }

        // Calculate rotation
        let firstVector = translatedTrajectory[index]
        let angle = atan2(firstVector.y, firstVector.x)

        // Apply rotation con el label correcto 'by'
        let alignedTrajectory = translatedTrajectory.map { $0.rotate(by: -angle) }

        return alignedTrajectory
    }


    func computeTotalDisplacement(coordinates: [LocationCoordinates]) -> Double {
        guard coordinates.count >= 2 else { return 0.0 }
        return coordinates.first!.distanceTo(coordinates.last!)
    }


    func estimateArConf() -> Double {
        let requiredPositions = 10
        let maxConfidence = 1.0
        var numOkPositions = 0

        var confidence = maxConfidence
        for i in stride(from: arLocationBuffer.count - 1, through: max(arLocationBuffer.count - requiredPositions, 0), by: -1) {
            if (arLocationBuffer[i].x == 0.0 && arLocationBuffer[i].y == 0.0) ||
                i < 1 ||
                (arLocationBuffer[i].y == arLocationBuffer[i - 1].y && arLocationBuffer[i].x == arLocationBuffer[i - 1].x) {
                break
            } else {
                numOkPositions += 1
            }
        }
        confidence = (Double(numOkPositions) / Double(requiredPositions)) * maxConfidence
        return confidence
    }

    func estimateSitumConf() -> Double {
        let requiredPositions = 10
        let maxConfidence = 1.0
        var numOkPositions = 0

        var confidence = maxConfidence
        for i in stride(from: situmLocationBuffer.count - 1, through: max(situmLocationBuffer.count - requiredPositions, 0), by: -1) {
            if situmLocationBuffer[i].accuracy > 5 && !situmLocationBuffer[i].hasBearing || i < 0 {
                break
            } else {
                numOkPositions += 1
            }
        }
        confidence = (Double(numOkPositions) / Double(requiredPositions)) * maxConfidence
        return confidence
    }

    func totalDisplacementConf(distance: Double) -> Double {
        let minDistanceThreshold = 10.0
        return distance > minDistanceThreshold ? 1.0 : distance / minDistanceThreshold
    }

    func odometriesDifferenceConf(difference: Double) -> Double {
        let diffThreshold = 10.0
        return difference > diffThreshold ? 0.0 : 1.0 - difference / diffThreshold
    }

    func checkIfHasToRefreshAndUpdateThreshold(conf: Double, arConf: Double, situmConf: Double) -> Bool {
        let currentTimestamp = Date().timeIntervalSince1970 * 1000

        if arConf < 0.8 || situmConf < 0.8 {
            resetThreshold()
            return true
        }
print("CONSTANT_QUALITY_DECREASE_RATE:   ", CONSTANT_QUALITY_DECREASE_RATE)
        if currentRefreshThreshold.value > 0.20 && currentTimestamp - currentRefreshThreshold.timestamp > 1000 {
            currentRefreshThreshold.value -= CONSTANT_QUALITY_DECREASE_RATE
        }

        if conf > currentRefreshThreshold.value + 0.2 {
            currentRefreshThreshold.value = conf
            currentRefreshThreshold.timestamp = currentTimestamp
            dynamicRefreshThreshold = currentRefreshThreshold
            return true
        } else if conf < currentRefreshThreshold.value && currentTimestamp - currentRefreshThreshold.timestamp > 1000 && currentRefreshThreshold.value > 0.20 {
            currentRefreshThreshold.value -= QUALITY_THRESHOLD_DECREASE_RATE
            currentRefreshThreshold.timestamp = currentTimestamp
            dynamicRefreshThreshold = currentRefreshThreshold
            return false
        } else if currentTimestamp - currentRefreshThreshold.timestamp > 30000 {
            currentRefreshThreshold.value = max(conf, 0.2)
            currentRefreshThreshold.timestamp = currentTimestamp
            dynamicRefreshThreshold = currentRefreshThreshold
            return true
        }

        return false
    }

    private func resetThreshold() {
        currentRefreshThreshold = RefreshThreshold(value: 0.2, timestamp: Date().timeIntervalSince1970 * 1000)
        dynamicRefreshThreshold = currentRefreshThreshold
    }
    
    func getInfoParameters() -> [String: Double]{
        
        var infoDebug: [String: Double] = [
            "globalQuality": quality,
            "DynamicRefreshThreshold": Double(dynamicRefreshThreshold.value),
            "arConf": arConf,
            "situmConf" : situmConf
        
        ]
        
        return infoDebug
        
    }
    
    
    
}

